pkg/controller/machine.go

```diff
@@ -266,14 +266,39 @@ func (c *controller) getMachineFromNode(nodeName string) (*v1alpha1.Machine, err
 }
 
 func (c *controller) updateMachineState(machine *v1alpha1.Machine) (*v1alpha1.Machine, error) {
+	nodeName := machine.Status.Node
 
-	if machine.Status.Node == "" {
-		// There are no objects mapped to this machine object
-		// Hence node status need not be propogated to machine object
-		return machine, nil
+	if nodeName == "" {
+		// Check if any existing node-object can be adopted.
+		nodeList, err := c.nodeLister.List(labels.Everything())
+		if err != nil {
+			klog.Errorf("Could not list the nodes due to error: %v", err)
+			return machine, err
+		}
+		for _, node := range nodeList {
+			nID, mID := decodeMachineID(node.Spec.ProviderID), decodeMachineID(machine.Spec.ProviderID)
```

Comment:
Is this required? I think comparing both the complete providerIDs should be enough right?

Reply:
>Is this required? I think comparing both the complete providerIDs should be enough right?

Yes, this seems to be required as `ProviderID` at machine object and node-object have slightly different formatting based on providers. And content is not exactly the same, but the last part of the IDs are always uuid and same. 

> we don't need to make the call decodeMachineID(machine.Spec.ProviderID) in every iteration.
This can be done just once before starting the iterations

Node-call has to be inside and machine-call could be outside the loop. Considering this gets executed only in rare cases when node-object-name is missing, I kept both of them inside for better readability, hope that's fine.

Reply:
> Yes, this seems to be required as ProviderID at machine object and node-object have slightly different formatting based on providers. And content is not exactly the same, but the last part of the IDs are always uuid and same.

When a machine is trying to adopt a node, the machine and node should belong to the same provider and the complete providerID should match right? This decode might not be required right?

Reply:
If it's not the case, can you provide me an example where you have seen differently?

Reply:
Yes, of course, they are the same providers. But I observed changes in the ProviderIDs at machine-object and node-objects, for instance, 
- aws:///eu-central-1/i-0123442 - at machine object.
- aws:///eu-central-1a/i-0123442 - at node object. 

This can happen, as at machine we calculate the providerID via driver-implementation but providerID on node-object comes from the kubelet. We should probably use the same kubelet function everywhere.

Reply:
Okay if this is the case. The current implementation makes sense. However, we need to update our driver create to return the right parameter then as per the kubelet.